dhall-lsp-server/src/Dhall/LSP/Backend/Parsing.hs

```diff
@@ -0,0 +1,54 @@
+module Dhall.LSP.Backend.Parsing (getLetInner, getLetAnnot) where
+
+import Dhall.Src (Src(..))
+import Dhall.Parser
+-- import Dhall.Parser.Combinators
+import Dhall.Parser.Token
+import Dhall.Parser.Expression
+
+import Control.Applicative (optional)
+import qualified Text.Megaparsec as Megaparsec
+import Text.Megaparsec (SourcePos(..))
+
+
+-- | Parse the outermost binding in a Src descriptor of a let-block and return
+--   the rest. Ex. on input `let a = 0 let b = a in b` parses `let a = 0 ` and
+--   returns the Src descriptor containing `let b = a in b`.
+getLetInner :: Src -> Maybe Src
+getLetInner (Src left _ text) = Megaparsec.parseMaybe (unParser parseLetInnerOffset) text
+ where parseLetInnerOffset = do
+          setSourcePos left
+          _let
```

Comment:
Hmm. Currently the parser for let bindings is somewhere inside a where block in the definition of `completeExpression :: Parser a -> Parser (Expr Src a)`, and references `expression`, which is also only defined in that where block. It would be possible to expose a `binding :: Parser (Binding Src a)` if we unfold the whole `where` monster.